modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java

```diff
@@ -159,19 +171,25 @@ public T next() {
             } catch (InterruptedException | ExecutionException e) {
                 Throwable cause = e.getCause();
 
-                if (cause != null) {
-                    if (cause.getClass().equals(NodeStoppingException.class)) {
-                        throw new NoSuchElementException();
-                    } else {
-                        if (cause.getClass().equals(NoSuchElementException.class)) {
-                            throw (NoSuchElementException) cause;
-                        }
-                    }
+                if (cause instanceof NodeStoppingException) {
+                    throw new NoSuchElementException();
+                }
+
+                if (cause instanceof NoSuchElementException) {
+                    throw (NoSuchElementException) cause;
                 }
 
                 LOG.debug("Unable to evaluate cursor hasNext command", e);
 
-                throw new IgniteInternalException(e);
+                if (e.getCause() instanceof IgniteInternalException) {
+                    IgniteInternalException exception = (IgniteInternalException) e.getCause();
+                    throw new IgniteInternalException(exception.traceId(), exception.code(), exception);
```

Comment:
It's interesting, why don't we use `CURSOR_EXECUTION_ERR` here as a code? Original code will be preserved in the cause.

Reply:
Well, it makes sense to me. Will change.